waku/waku_api/message_cache.nim

```diff
@@ -15,68 +15,292 @@ import
+  messages: seq[WakuMessage]
+  pubsubTopics: seq[PubsubTopic]
+  contentTopics: seq[ContentTopic]
+  pubsubIndex: seq[(int, int)]  # Pubsub topic index mapped to message index
+  contentIndex: seq[(int, int)] # Content topic index mapped to message index
 
-type MessageCacheResult*[T] = Result[T, cstring]
+func `$`*(self: MessageCache): string =
+  "Messages: " & $self.messages.len &
+  " \nPubsubTopics: " & $self.pubsubTopics &
+  " \nContentTopics: " & $self.contentTopics &
+  " \nPubsubIndex: " & $self.pubsubIndex &
+  " \nContentIndex: " & $self.contentIndex
 
-type MessageCache*[K] = ref object
-  capacity: uint
-  table: Table[K, seq[WakuMessage]]
-
-func init*[K](T: type MessageCache[K], capacity=DefaultMessageCacheCapacity): T =
-  MessageCache[K](
-    capacity: capacity,
-    table: initTable[K, seq[WakuMessage]]()
+func init*(T: type MessageCache, capacity=DefaultMessageCacheCapacity): T =
+  MessageCache(
+    capacity: capacity
   )
 
+proc messagesCount*(self: MessageCache): int =
+  self.messages.len
 
-proc isSubscribed*[K](t: MessageCache[K], topic: K): bool =
-  t.table.hasKey(topic)
+proc pubsubTopicCount*(self: MessageCache): int =
+  self.pubsubTopics.len
 
-proc subscribe*[K](t: MessageCache[K], topic: K) =
-  if t.isSubscribed(topic):
-    return
-  t.table[topic] = @[]
+proc contentTopicCount*(self: MessageCache): int =
+  self.contentTopics.len
 
-proc unsubscribe*[K](t: MessageCache[K], topic: K) =
-  if not t.isSubscribed(topic):
-    return
-  t.table.del(topic)
+proc pubsubSearch(self: MessageCache, pubsubTopic: PubsubTopic): Option[int] =
+  # Return some with the index if found none otherwise.
+  
+  for i, topic in self.pubsubTopics:
+    if topic == pubsubTopic:
+      return some(i)
+
+  return none(int)
+
+proc contentSearch(self: MessageCache, contentTopic: ContentTopic): Option[int] =
+  # Return some with the index if found none otherwise.
+  
+  for i, topic in self.contentTopics:
+    if topic == contentTopic:
+      return some(i)
+
+  return none(int)
+
+proc isPubsubSubscribed*(self: MessageCache, pubsubTopic: PubsubTopic): bool =
+  self.pubsubSearch(pubsubTopic).isSome()
+
+proc isContentSubscribed*(self: MessageCache, contentTopic: ContentTopic): bool =
+  self.contentSearch(contentTopic).isSome()
+
+proc pubsubSubscribe*(self: MessageCache, pubsubTopic: PubsubTopic) =
+  if self.pubsubSearch(pubsubTopic).isNone():
+    self.pubsubTopics.add(pubsubTopic)
+
+proc contentSubscribe*(self: MessageCache, contentTopic: ContentTopic) =
+  if self.contentSearch(contentTopic).isNone():
+    self.contentTopics.add(contentTopic)
+
+proc removeMessage(self: MessageCache, idx: int) =
+  let lastIndex = self.messages.high
+  self.messages.del(idx)
+  
+  # update indexes
+  var j = self.pubsubIndex.high
+  while j > -1:
+    let (pId, mId) = self.pubsubIndex[j]
+
+    if mId == idx:
+      self.pubsubIndex.del(j)
+    elif mId == lastIndex:
+      self.pubsubIndex[j] = (pId, idx)
```

Comment:
what are we doing here? not sure I understand this special case of the last index

Reply:
`del` is a swap to last index. So we need to also update `pubsubIndex` to point to the message after the swap.

`messages[last]` swap place with `messages[idx]`